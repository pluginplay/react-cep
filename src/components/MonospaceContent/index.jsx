import React from 'react'
import { FlexItem } from 'styled-flex-component'
import styled from 'styled-components'

class MonospaceContent extends React.Component {

  onClick (e) {
    e.preventDefault()
    const range = document.createRange()
    const sel = window.getSelection()
    range.selectNodeContents(e.target)
    sel.removeAllRanges()
    sel.addRange(range)
  }

  render () {
    return (
      <FlexItem {...this.props} onClick={this.onClick}>
        {this.props.children}
      </FlexItem>
    )
  }

}

export default styled(MonospaceContent)`
  background: ${props => props.theme.background.darker};
  border-radius: 2px;
  padding: 5px;
  margin: 0;
  width: 100%;
  box-sizing: border-box;
  font-family: monospace;
  overflow: auto;
  text-align: left;
`
