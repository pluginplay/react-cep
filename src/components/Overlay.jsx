import Flex from 'styled-flex-component'

export default Flex.extend`
  transition: opacity 0.25s linear, z-index 0s linear ${props => props.visible ? '' : '0.25s'};
  text-align: center;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background: ${props => props.theme.background.dark};
  z-index: ${props => props.visible ? 3 : -1};
  opacity: ${props => props.visible ? 1 : 0};
  a {
    color: white;
    text-decoration: none;
    opacity: 0.75;
    transition: opacity 0.2s linear;
    &:hover {
      opacity: 1;
    }
  }
`
