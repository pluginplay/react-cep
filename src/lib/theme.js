import * as csInterface from 'cep-lib/csinterface'
import { lighten, darken } from 'polished'

export const getDefaultTheme = () => {
  const ae = csInterface.getThemeInformation()
  ae.systemHighlightColor = '#0078d7' // We have to hard-code this because Windows gets the wrong color.
  return {
    ae,
    background: {
      lighter: lighten(0.2, ae.panelBackgroundColor),
      light: lighten(0.1, ae.panelBackgroundColor),
      dark: darken(0.05, ae.panelBackgroundColor),
      darker: darken(0.1, ae.panelBackgroundColor)
    },
    colors: {
      green: '#23EB87',
      red: '#eb093f',
      blue: ae.systemHighlightColor
    }
  }
}
