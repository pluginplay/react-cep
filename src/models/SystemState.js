import { observable, action } from 'mobx'

export default class SystemState {
  @observable progress = 0
  @observable progressLabel = 'Loading'
  @observable showProgress = false
  @observable showProgressComplete = false
  @observable errorMessage = null
  @observable errorObject = null
  @observable showErrorOverlay = false
  @observable debugMessage = null
  @observable showDebugOverlay = null
  @observable confirmationMessage = null
  @observable confirmationCallback = null
  @observable showConfirmationOverlay = false

  @action setProgress (progress, label, showProgress = true) {
    this.showProgressComplete = false
    this.showProgress = showProgress
    this.progressLabel = label
    this.progress = progress
  }

  @action progressComplete () {
    this.progress = 100
    this.showProgressComplete = true
    setTimeout(() => {
      this.hideProgress()
    }, 3000)
  }

  @action hideProgress () {
    this.showProgress = false
  }

  @action setErrorMessage (errorText, error) {
    this.errorMessage = errorText
    this.errorObject = error
    this.showErrorOverlay = true
  }

  @action hideErrorMessage () {
    this.errorMessage = ''
    this.errorObject = null
    this.showErrorOverlay = false
  }

  @action addDebugMessage (message) {
    console.log(`[DEBUG] ${message}`)
    if (this.debugMessage === null) {
      this.debugMessage = message
    } else {
      this.debugMessage += `\n${message}`
    }
  }

  @action showDebug () {
    this.showDebugOverlay = true
  }

  @action clearDebug () {
    this.showDebugOverlay = false
    this.debugMessage = null
  }

  @action showConfirmation (message, callback) {
    this.showConfirmationOverlay = true
    this.confirmationMessage = message
    this.confirmationCallback = callback
  }

  @action hideConfirmation () {
    this.showConfirmationOverlay = false
    this.confirmationMessage = null
    this.confirmationCallback = null
  }
}
