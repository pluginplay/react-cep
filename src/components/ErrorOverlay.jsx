import React from 'react'
import { observer } from 'mobx-react'
import PropTypes from 'prop-types'
import Overlay from './Overlay'
import OverlayContent from './OverlayContent'
import ColorHeader from './ColorHeader'
import MonospaceContent from './MonospaceContent'
import styled from 'styled-components'

@observer
class ErrorOverlay extends React.Component {
  constructor (props) {
    super(props)
    this._onCloseClickedBound = this.onCloseClicked.bind(this)
    this._onWhatMeanClickedBound = this.onWhatMeanClicked.bind(this)
    this._onReportClickedBound = this.onReportClicked.bind(this)
  }

  onWhatMeanClicked (e) {
    if (e) {
      e.preventDefault()
    }
    this.props.systemState.showConfirmation('When you send the developers an error report, you send very useful ' +
      'information about what went wrong that allows us to improve the plugin. The error report will contain ' +
      'your AE version, installed extensions, and what operating system you\'re running.')
  }

  onReportClicked (e) {
    if (e) {
      e.preventDefault()
    }
    if (this.props.onReportClicked) {
      this.props.onReportClicked()
    }
    this.props.systemState.hideErrorMessage()
  }

  onCloseClicked () {
    this.props.systemState.hideErrorMessage()
  }

  render () {
    return (
      <Overlay visible={this.props.systemState.showErrorOverlay} column full className={this.props.className}>
        <ColorHeader color={'red'}>
          <span>Oh dear, an error!</span>
          <a href={'#'} onClick={this._onCloseClickedBound}>Close</a>
        </ColorHeader>
        <OverlayContent full column>
          <p>
            Looks like we've run into a problem. You could help the developers out and <a
              href={'#'} className={'send-button'} onClick={this._onReportClickedBound}>send the error report.</a>
            <br /><a href={'#'} onClick={this._onWhatMeanClickedBound}>What does this mean?</a>
          </p>
          <MonospaceContent grow={1}>{this.props.systemState.errorMessage}</MonospaceContent>
        </OverlayContent>
      </Overlay>
    )
  }
}

ErrorOverlay.propTypes = {
  systemState: PropTypes.object.isRequired,
  onReportClicked: PropTypes.func
}

export default styled(ErrorOverlay)`
  .send-button {
    background: ${props => props.theme.ae.systemHighlightColor};
    padding: 1px 2px;
    color: white;
    font-weight: bold;
    white-space: nowrap;
  }
`
