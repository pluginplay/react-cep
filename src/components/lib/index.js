import SystemState from '../../models/SystemState'
import PersistedState from '../../models/PersistedState'

export { SystemState, PersistedState }
