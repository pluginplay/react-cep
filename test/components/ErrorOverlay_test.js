import React from 'react'
import { shallow } from 'enzyme'
import { expect, withTheme } from '../setup'
import { getDefaultTheme } from '../../src/lib/theme'
import renderer from 'react-test-renderer'
import SystemState from '../../src/models/SystemState'
import ErrorOverlay from '../../src/components/ErrorOverlay'

import Overlay from '../../src/components/Overlay'
import MonospaceContent from '../../src/components/MonospaceContent'

let systemState = null
let showConfirmationSpy = null
let hideErrorMessageSpy = null

beforeEach(() => {
  showConfirmationSpy = jest.spyOn(SystemState.prototype, 'showConfirmation')
  hideErrorMessageSpy = jest.spyOn(SystemState.prototype, 'hideErrorMessage')
  systemState = new SystemState()
  systemState.showErrorOverlay = true
  systemState.errorMessage = 'foobar'
})

afterEach(() => {
  showConfirmationSpy.mockRestore()
  hideErrorMessageSpy.mockRestore()
})

const getWrapper = (props = {}) => {
  return shallow(<ErrorOverlay theme={getDefaultTheme()} systemState={systemState} {...props} />).dive()
}

it('is only visible when showErrorOverlay is true', () => {
  const wrapper = getWrapper()
  expect(wrapper.find(Overlay).first()).to.have.prop('visible').equal(true)
})

it('hides the error message when close is clicked', () => {
  const wrapper = getWrapper()
  wrapper.find('a').findWhere(a => a.text() === 'Close').first().simulate('click')
  jestExpect(hideErrorMessageSpy).toHaveBeenCalledTimes(1)
})

it('displays an explanation when "What does this mean?" is clicked', () => {
  const wrapper = getWrapper()
  wrapper.find('a').findWhere(a => a.text() === 'What does this mean?').first().simulate('click')
  expect(showConfirmationSpy.mock.calls).to.have.length(1)
  expect(showConfirmationSpy.mock.calls[0][0]).to.be.a('string')
})

context('when report is clicked', () => {
  it('hides the error message', () => {
    const wrapper = getWrapper()
    wrapper.find('a.send-button').first().simulate('click')
    jestExpect(hideErrorMessageSpy).toHaveBeenCalledTimes(1)
  })

  it('calls the onReportClicked callback if it exists', () => {
    const onReportClicked = jest.fn()
    const wrapper = getWrapper({ onReportClicked })
    wrapper.find('a.send-button').first().simulate('click')
    jestExpect(onReportClicked).toHaveBeenCalledTimes(1)
  })
})

it('displays the proper error message', () => {
  const wrapper = getWrapper()
  expect(wrapper.find(MonospaceContent).first().children().first()).to.have.text('foobar')
})

it('renders the proper HTML structure', () => {
  const tree = renderer
    .create(withTheme(<ErrorOverlay systemState={systemState} />))
    .toJSON()
  jestExpect(tree).toMatchSnapshot()
})
