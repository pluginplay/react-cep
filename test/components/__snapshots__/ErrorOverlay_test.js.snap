// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the proper HTML structure 1`] = `
.c5 {
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
  -webkit-flex-basis: auto;
  -ms-flex-preferred-size: auto;
  flex-basis: auto;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  display: block;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  height: 100%;
  -webkit-flex-basis: 100%;
  -ms-flex-preferred-size: 100%;
  flex-basis: 100%;
  -webkit-transition: opacity 0.25s linear,z-index 0s linear;
  transition: opacity 0.25s linear,z-index 0s linear;
  text-align: center;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background: #191919;
  z-index: 3;
  opacity: 1;
}

.c1 a {
  color: white;
  -webkit-text-decoration: none;
  text-decoration: none;
  opacity: 0.75;
  -webkit-transition: opacity 0.2s linear;
  transition: opacity 0.2s linear;
}

.c1 a:hover {
  opacity: 1;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  height: 100%;
  -webkit-flex-basis: 100%;
  -ms-flex-preferred-size: 100%;
  flex-basis: 100%;
  padding: 10px;
  box-sizing: border-box;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  background: #eb093f;
  color: white;
  padding: 10px;
  font-size: 16px;
  font-weight: bold;
  line-height: 1;
}

.c2 span {
  margin: 0 auto;
}

.c2 a {
  margin: 0 3px 0 auto;
  font-size: 14px;
  font-weight: normal;
  color: white;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c4 {
  background: #0c0c0c;
  border-radius: 2px;
  padding: 5px;
  margin: 0;
  width: 100%;
  box-sizing: border-box;
  font-family: monospace;
  overflow: auto;
  text-align: left;
}

.c0 .send-button {
  background: #0078d7;
  padding: 1px 2px;
  color: white;
  font-weight: bold;
  white-space: nowrap;
}

<div
  className="c0 c1"
>
  <div
    className="c2"
    color="red"
  >
    <span>
      Oh dear, an error!
    </span>
    <a
      href="#"
      onClick={[Function]}
    >
      Close
    </a>
  </div>
  <div
    className="c3"
  >
    <p>
      Looks like we've run into a problem. You could help the developers out and 
      <a
        className="send-button"
        href="#"
        onClick={[Function]}
      >
        send the error report.
      </a>
      <br />
      <a
        href="#"
        onClick={[Function]}
      >
        What does this mean?
      </a>
    </p>
    <div
      className="c4 c5"
      onClick={[Function]}
    >
      foobar
    </div>
  </div>
</div>
`;
